### PR TITLE
Inject Dependencies using Singleton Pattern

### PR DESCRIPTION
**Pull request by jonajonlee**
### Example

_in some statically required module on the app using this_

``` javascript
const Yaldic = require('yaldic')
const Domain = require('kuss/lib/domain')
const Elasticsearch = require('elasticsearch')
const Mysql2 = require('mysql2')
const MysqlStore = require('kuss/lib/mysql')
const domainPath = 'user/home/homeboy/social-hoarder/domain'

const CONFIG = require('../../config')

const myYaldic = Yaldic({
  DO: () => Domain(domainPath)
, DAO: () => MysqlStore(Mysql2.createPool(CONFIG.DATABASE.MYSQL))
, elasticsearch: () => Elasticsearch.client(CONFIG.ELASTICSEARCH)
})
```

_in a file that uses the statically required module on the same app_

``` javascript
const Deps = require('../../lib/container')
const Handler = require('../../handler')

const doSomethingCool = (id) => {
  const DO = Deps.get('DO')

  return DO.Archive.getById(id)
}

module.exports = { doSomethingCool }
```

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,99 +1,50 @@
...
+const throwErrorIfNoSpec = R.when(
+  R.either(R.isEmpty, R.isNil)
+, () => { throw new Error('Spec object is required') }
+)
 
-function validateNode(node) {
 
-  if (typeof node === 'string')
-    throw new TypeError('Dependency node cannot be a string');
+// const throwNotInSpec = (key) => {
```
Comment (scull7) [inline index.js]: Did you mean to leave this here?

File: index.js [hunk 2]
```diff
@@ -1,99 +1,50 @@
...
 
-}
-
-function Yaldic({ allow_overwrite = false } = {}) {
-
-  const graph = new DepGraph();
-
-
-  return {
+const throwErrorIfNotInSpec = (spec) => R.when(
```
Comment (scull7) [inline index.js:33]: You should use the [props-check](https://www.npmjs.com/package/props-check) module to output an error when the user asks for something not in the spec.

**Review by scull7 — CHANGES_REQUESTED** [review]

File: index.js (hunk 2 above)
Comment (jonajonlee) [reply index.js:33]: added propsCheck

File: index.js (hunk 1 above)
Comment (jonajonlee) [reply index.js]: removed this

**Review by scull7 — APPROVED** [review]

**Review by Kolossion — APPROVED** [review]
:shipit:

**Review by mattsmee — APPROVED** [review]
:shipit:

**Comment by scull7** [comment]
You didn't cover the DNE error case.  

screenshot 2016-10-25 13 55 15

**Comment by jonajonlee** [comment]
@scull7 

One Hundo Percento

image

**Pull request merged**